src/coreclr/jit/flowgraph.cpp

```diff
@@ -436,6 +436,338 @@ BasicBlock* Compiler::fgCreateGCPoll(GCPollType pollType, BasicBlock* block)
+        // This leads to some size savings on NativeAOT
+        if ((staticBaseAddr.addr == flagAddr.addr) && (staticBaseAddr.accessType == flagAddr.accessType))
+        {
+            cachedStaticBase = fgInsertCommaFormTemp(&baseAddr);
+        }
+
+        // Don't fold ADD(CNS1, CNS2) here since the result won't be reloc-friendly for AOT
+        GenTree* offsetNode     = gtNewOperNode(GT_ADD, TYP_I_IMPL, baseAddr, gtNewIconNode(isInitOffset));
+        isInitedActualValueNode = gtNewIndir(TYP_I_IMPL, offsetNode, GTF_IND_NONFAULTING);
+        isInitedActualValueNode->gtFlags |= GTF_GLOB_REF;
+
+        // 0 means "initialized" on NativeAOT
+        isInitedExpectedValue = gtNewIconNode(0, TYP_I_IMPL);
+    }
+    else
+    {
+        assert(isInitOffset == 0);
+
+        isInitedActualValueNode = gtNewIndOfIconHandleNode(TYP_INT, (size_t)flagAddr.addr, GTF_ICON_GLOBAL_PTR, false);
+
+        // Check ClassInitFlags::INITIALIZED_FLAG bit
+        isInitedActualValueNode = gtNewOperNode(GT_AND, TYP_INT, isInitedActualValueNode, gtNewIconNode(1));
+        isInitedExpectedValue   = gtNewIconNode(1);
+    }
+
+    GenTree* isInitedCmp = gtNewOperNode(GT_EQ, TYP_INT, isInitedActualValueNode, isInitedExpectedValue);
+    isInitedCmp->gtFlags |= GTF_RELOP_JMP_USED;
+    BasicBlock* isInitedBb =
+        fgNewBBFromTreeAfter(BBJ_COND, prevBb, gtNewOperNode(GT_JTRUE, TYP_VOID, isInitedCmp), debugInfo);
+
+    // Fallback basic block
+    // TODO-CQ: for JIT we can replace the original call with CORINFO_HELP_INITCLASS
+    // that only accepts a single argument
+    BasicBlock* helperCallBb = fgNewBBFromTreeAfter(BBJ_NONE, isInitedBb, call, debugInfo, true);
+
+    GenTree* replacementNode = nullptr;
+    if (retValKind == SHRV_STATIC_BASE_PTR)
+    {
+        // Replace the call with a constant pointer to the statics base
+        assert(staticBaseAddr.addr != nullptr);
+
+        // Use local if the addressed is already materialized and cached
+        if (cachedStaticBase != nullptr)
+        {
+            assert(staticBaseAddr.accessType == IAT_VALUE);
+            replacementNode = cachedStaticBase;
+        }
+        else if (staticBaseAddr.accessType == IAT_VALUE)
+        {
+            replacementNode = gtNewIconHandleNode((size_t)staticBaseAddr.addr, GTF_ICON_STATIC_HDL);
+        }
+        else
+        {
+            assert(staticBaseAddr.accessType == IAT_PVALUE);
+            replacementNode =
+                gtNewIndOfIconHandleNode(TYP_I_IMPL, (size_t)staticBaseAddr.addr, GTF_ICON_GLOBAL_PTR, false);
+        }
+    }
+
+    if (replacementNode == nullptr)
+    {
+        (*callUse)->gtBashToNOP();
+    }
+    else
+    {
+        *callUse = replacementNode;
+    }
+
+    fgMorphStmtBlockOps(block, stmt);
+    gtUpdateStmtSideEffects(stmt);
+
+    //
+    // Update preds in all new blocks
+    //
+
+    // Unlink block and prevBb
+    fgRemoveRefPred(block, prevBb);
```

Comment:
would be good to have a basic block layout in comments to understand how blocks are arranged.

Reply:
Ok will add

Reply:
Added